metagraph/src/graph/representation/canonical_dbg.cpp

```diff
@@ -191,6 +223,64 @@ ::call_outgoing_kmers(node_index node, const OutgoingEdgeCallback &callback) con
     }
 }
 
+void CanonicalDBG
+::get_kmers_from_prefix(node_index node, std::vector<node_index> &parents) const {
+    const auto &alphabet = graph_.alphabet();
+    std::string rev_seq = std::string(1, '\0') + get_node_sequence(node).substr(0, get_k() - 1);
+    ::reverse_complement(rev_seq.begin(), rev_seq.end());
+    assert(rev_seq.back() == '\0');
+
+    const auto *dbg_succ = dynamic_cast<const DBGSuccinct*>(&graph_);
+    if (dbg_succ) {
+        const auto &boss = dbg_succ->get_boss();
+        auto encoded = boss.encode(std::string_view(rev_seq.data(), get_k() - 1));
+
+        auto [edge, edge_2, end] = boss.index_range(encoded.begin(), encoded.end());
+
+        if (edge && edge_2 && end == encoded.end()) {
+            assert(edge == edge_2);
+            do {
+                node_index prev = dbg_succ->boss_to_kmer_index(edge);
+                if (prev) {
+                    char c = alphabet[boss.get_W(edge) % boss.alph_size];
+                    if (c == boss::BOSS::kSentinel)
+                        continue;
+
+                    ::reverse_complement(&c, &c + 1);
+                    size_t i = boss.encode(c);
+
+                    if (parents[i] == DeBruijnGraph::npos) {
+                        if (!primary_)
+                            rev_comp_cache_.Put(prev, prev + offset_);
+
+                        parents[i] = prev + offset_;
+                    }
+                }
+
+            } while (--edge && !boss.get_last(edge));
```

Comment:
I think there should be a routine like `call_outgoing` in BOSS for this do-while loop.

Reply:
There isn't, so I added one